In [43]:
%pip install -q nltk
%pip install -q pymorphy2
%pip install -q spacy
# %pip install -q spacy download ru_core_news_lg
%pip install https://github.com/explosion/spacy-models/releases/download/ru_core_news_sm-3.7.0/ru_core_news_sm-3.7.0-py3-none-any.whl
%pip install https://github.com/explosion/spacy-models/releases/download/ru_core_news_lg-3.7.0/ru_core_news_lg-3.7.0-py3-none-any.whl

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\velic\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\velic\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\velic\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\velic\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.
You should consider upgrading via the 'c:\Users\velic\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\velic\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


In [44]:
import pymorphy2
import re
import spacy

In [45]:
with open('./krestyanskie_deti1861.txt', 'r', encoding='utf-8') as file:
    book1861 = file.read()

with open('./moroz_krasnyy_nos_1864.txt', 'r', encoding='ansi') as file:
    book1864 = file.read()

with open('./zheleznaya_doroga_1865.txt', 'r', encoding='ansi') as file:
    book1865 = file.read()

with open('./dedushka_mazay_i_zaycy_1871.txt', 'r', encoding='ansi') as file:
    book1871 = file.read()

with open('./russkie_zhenshiny_1872.txt', 'r', encoding='utf-8') as file:
    book1872 = file.read()

with open('./komu_na_rusi_zhit_horosho_1874.txt', 'r', encoding='utf-8') as file:
    book1874 = file.read()

In [46]:
def word_tokenizer(book1861):
    words = re.findall(r'\w+', book1861)
    return words

print('Лексическое разнообразие произведения "Крестьянские дети":')
tokens = word_tokenizer(book1861)
wordsCount = len(tokens)
uniqueWords = list(set(tokens))
uniqueWordsCount = len(uniqueWords)
ttr1861 = uniqueWordsCount/wordsCount
print(ttr1861)


morph = pymorphy2.MorphAnalyzer()

tokens = word_tokenizer(book1861)

countADJ1861 = 0
for token in tokens:
    parsed_word = morph.parse(token)
    for parsed in parsed_word:
        if "ADJF" in parsed.tag:
            countADJ1861 += 1

print(f'Количество прилагательных в тексте: {countADJ1861}')

unique_adjectives = set()
for token in tokens:
    parsed_word = morph.parse(token)
    for parsed in parsed_word:
        if "ADJF" in parsed.tag:
            unique_adjectives.add(parsed.normal_form)

countUADJ1861 = len(unique_adjectives)
print(f'Количество уникальных прилагательных в тексте: {countUADJ1861}')

ttrADJ1861 = countUADJ1861/countADJ1861
print('Разнообразие прилагательных:', ttrADJ1861)


# nlp = spacy.load("ru_core_news_sm")

# doc = nlp(book1861)

# for sentence in doc.sents:
#     print("Предложение:", sentence.text)
#     for token in sentence:
#         print("Слово:", token.text)
#         print("Лемма:", token.lemma_)
#         print("Частеречный тег:", token.pos_)

Лексическое разнообразие произведения "Крестьянские дети":
0.6877243359655419


Количество прилагательных в тексте: 749
Количество уникальных прилагательных в тексте: 118
Разнообразие прилагательных: 0.157543391188251


In [47]:
def word_tokenizer(book1864):
    words = re.findall(r'\w+', book1864)
    return words

print('Лексическое разнообразие произведения "Мороз, Красный нос":')
tokens = word_tokenizer(book1864)
wordsCount = len(tokens)
uniqueWords = list(set(tokens))
uniqueWordsCount = len(uniqueWords)
ttr1864 = uniqueWordsCount/wordsCount
print(ttr1864)


morph = pymorphy2.MorphAnalyzer()

tokens = word_tokenizer(book1864)

countADJ1864 = 0
for token in tokens:
    parsed_word = morph.parse(token)
    for parsed in parsed_word:
        if "ADJF" in parsed.tag:
            countADJ1864 += 1

print(f'Количество прилагательных в тексте: {countADJ1864}')

unique_adjectives = set()
for token in tokens:
    parsed_word = morph.parse(token)
    for parsed in parsed_word:
        if "ADJF" in parsed.tag:
            unique_adjectives.add(parsed.normal_form)

countUADJ1864 = len(unique_adjectives)
print(f'Количество уникальных прилагательных в тексте: {countUADJ1864}')

ttrADJ1864 = countUADJ1864/countADJ1864
print('Разнообразие прилагательных:', ttrADJ1864)

Лексическое разнообразие произведения "Мороз, Красный нос":
0.5918367346938775
Количество прилагательных в тексте: 2559
Количество уникальных прилагательных в тексте: 251
Разнообразие прилагательных: 0.09808518952715904


In [48]:
def word_tokenizer(book1865):
    words = re.findall(r'\w+', book1865)
    return words

print('Лексическое разнообразие произведения "Железная дорога":')
tokens = word_tokenizer(book1865)
wordsCount = len(tokens)
uniqueWords = list(set(tokens))
uniqueWordsCount = len(uniqueWords)
ttr1865 = uniqueWordsCount/wordsCount
print(ttr1865)


morph = pymorphy2.MorphAnalyzer()

tokens = word_tokenizer(book1865)

countADJ1865 = 0
for token in tokens:
    parsed_word = morph.parse(token)
    for parsed in parsed_word:
        if "ADJF" in parsed.tag:
            countADJ1865 += 1

print(f'Количество прилагательных в тексте: {countADJ1865}')

unique_adjectives = set()
for token in tokens:
    parsed_word = morph.parse(token)
    for parsed in parsed_word:
        if "ADJF" in parsed.tag:
            unique_adjectives.add(parsed.normal_form)

countUADJ1865 = len(unique_adjectives)
print(f'Количество уникальных прилагательных в тексте: {countUADJ1865}')

ttrADJ1865 = countUADJ1865/countADJ1865
print('Разнообразие прилагательных:', ttrADJ1865)

Лексическое разнообразие произведения "Железная дорога":
0.7598944591029023
Количество прилагательных в тексте: 336
Количество уникальных прилагательных в тексте: 79
Разнообразие прилагательных: 0.23511904761904762


In [49]:
def word_tokenizer(book1871):
    words = re.findall(r'\w+', book1871)
    return words

print('Лексическое разнообразие произведения "Дедушка Мазай и зайцы":')
tokens = word_tokenizer(book1871)
wordsCount = len(tokens)
uniqueWords = list(set(tokens))
uniqueWordsCount = len(uniqueWords)
ttr1871 = uniqueWordsCount/wordsCount
print(ttr1871)


morph = pymorphy2.MorphAnalyzer()

tokens = word_tokenizer(book1871)

countADJ1871 = 0
for token in tokens:
    parsed_word = morph.parse(token)
    for parsed in parsed_word:
        if "ADJF" in parsed.tag:
            countADJ1871 += 1

print(f'Количество прилагательных в тексте: {countADJ1871}')

unique_adjectives = set()
for token in tokens:
    parsed_word = morph.parse(token)
    for parsed in parsed_word:
        if "ADJF" in parsed.tag:
            unique_adjectives.add(parsed.normal_form)

countUADJ1871 = len(unique_adjectives)
print(f'Количество уникальных прилагательных в тексте: {countUADJ1871}')

ttrADJ1871 = countUADJ1871/countADJ1871
print('Разнообразие прилагательных:', ttrADJ1871)

Лексическое разнообразие произведения "Дедушка Мазай и зайцы":
0.7258485639686684
Количество прилагательных в тексте: 639
Количество уникальных прилагательных в тексте: 48
Разнообразие прилагательных: 0.07511737089201878


In [50]:
def word_tokenizer(book1872):
    words = re.findall(r'\w+', book1872)
    return words

print('Лексическое разнообразие произведения "Русские женщины":')
tokens = word_tokenizer(book1872)
wordsCount = len(tokens)
uniqueWords = list(set(tokens))
uniqueWordsCount = len(uniqueWords)
ttr1872 = uniqueWordsCount/wordsCount
print(ttr1872)


morph = pymorphy2.MorphAnalyzer()

tokens = word_tokenizer(book1872)

countADJ1872 = 0
for token in tokens:
    parsed_word = morph.parse(token)
    for parsed in parsed_word:
        if "ADJF" in parsed.tag:
            countADJ1872 += 1

print(f'Количество прилагательных в тексте: {countADJ1872}')

unique_adjectives = set()
for token in tokens:
    parsed_word = morph.parse(token)
    for parsed in parsed_word:
        if "ADJF" in parsed.tag:
            unique_adjectives.add(parsed.normal_form)

countUADJ1872 = len(unique_adjectives)
print(f'Количество уникальных прилагательных в тексте: {countUADJ1872}')

ttrADJ1872 = countUADJ1872/countADJ1872
print('Разнообразие прилагательных:', ttrADJ1872)

Лексическое разнообразие произведения "Русские женщины":
0.45311400422237863
Количество прилагательных в тексте: 6903
Количество уникальных прилагательных в тексте: 451
Разнообразие прилагательных: 0.06533391279153991


In [51]:
def word_tokenizer(book1874):
    words = re.findall(r'\w+', book1874)
    return words

print('Лексическое разнообразие произведения "Кому на Руси жить хорошо":')
tokens = word_tokenizer(book1874)
wordsCount = len(tokens)
uniqueWords = list(set(tokens))
uniqueWordsCount = len(uniqueWords)
ttr1874 = uniqueWordsCount/wordsCount
print(ttr1874)


morph = pymorphy2.MorphAnalyzer()

tokens = word_tokenizer(book1874)

countADJ1874 = 0
for token in tokens:
    parsed_word = morph.parse(token)
    for parsed in parsed_word:
        if "ADJF" in parsed.tag:
            countADJ1874 += 1

print(f'Количество прилагательных в тексте: {countADJ1874}')

unique_adjectives = set()
for token in tokens:
    parsed_word = morph.parse(token)
    for parsed in parsed_word:
        if "ADJF" in parsed.tag:
            unique_adjectives.add(parsed.normal_form)

countUADJ1874 = len(unique_adjectives)
print(f'Количество уникальных прилагательных в тексте: {countUADJ1874}')

ttrADJ1874 = countUADJ1874/countADJ1874
print('Разнообразие прилагательных:', ttrADJ1874)

Лексическое разнообразие произведения "Кому на Руси жить хорошо":
0.40199765706886986
Количество прилагательных в тексте: 15403
Количество уникальных прилагательных в тексте: 1120
Разнообразие прилагательных: 0.07271310783613581


In [52]:
import pandas as pd

data = {
    'Text': ['book1861', 'book1864', 'book1865', 'book1871', 'book1872', 'book1874'],
    'Разнообразие прилагательных': [ttrADJ1861, ttrADJ1864, ttrADJ1865, ttrADJ1871, ttrADJ1872, ttrADJ1874],
    'Лексическое разнообразие слов': [ttr1861, ttr1864, ttr1865, ttr1871, ttr1872, ttr1874]
}

df = pd.DataFrame(data)

def apply_styles_to_dataframe(dataframe):
    return (
        dataframe.style
        .set_properties(**{'background-color': 'pink', 'color': 'black'})
        .set_table_styles([
            {'selector': 'th', 'props': [('background-color', 'lightblue'), ('color', 'black'), ('font-weight', 'bold')]},
        ])
    )

apply_styles_to_dataframe(df)

In [53]:
from collections import Counter

nlp = spacy.load("ru_core_news_lg")


doc1861 = nlp(book1861)
tokens1 = [token.text for token in doc1861 if token.is_alpha and token.pos_ not in ["ADP", "CONJ", "PART", "SCONJ", "ADV", "PRON"] and token.text.isalpha() and len(token.text) > 1]

doc1864 = nlp(book1864)
tokens2 = [token.text for token in doc1864 if token.is_alpha and token.pos_ not in ["ADP", "CONJ", "PART", "SCONJ", "ADV", "PRON"] and token.text.isalpha() and len(token.text) > 1]

doc1865 = nlp(book1865)
tokens3 = [token.text for token in doc1865 if token.is_alpha and token.pos_ not in ["ADP", "CONJ", "PART", "SCONJ", "ADV", "PRON"] and token.text.isalpha() and len(token.text) > 1]

doc1871 = nlp(book1871)
tokens4 = [token.text for token in doc1871 if token.is_alpha and token.pos_ not in ["ADP", "CONJ", "PART", "SCONJ", "ADV", "PRON"] and token.text.isalpha() and len(token.text) > 1]

doc1872 = nlp(book1872)
tokens5 = [token.text for token in doc1872 if token.is_alpha and token.pos_ not in ["ADP", "CONJ", "PART", "SCONJ", "ADV", "PRON"] and token.text.isalpha() and len(token.text) > 1]

doc1874 = nlp(book1874)
tokens6 = [token.text for token in doc1874 if token.is_alpha and token.pos_ not in ["ADP", "CONJ", "PART", "SCONJ", "ADV", "PRON"] and token.text.isalpha() and len(token.text) > 1]

# Подсчет частоты слов в каждом тексте
word_freq1 = Counter(tokens1)
word_freq2 = Counter(tokens2)
word_freq3 = Counter(tokens3)
word_freq4 = Counter(tokens4)
word_freq5 = Counter(tokens5)
word_freq6 = Counter(tokens6)

print('Самые часто встречающиеся слова в тексте "Крестьянкие дети":')
for word, freq in word_freq1.most_common(5):
    token = nlp(word)[0]
    pos_tag = token.pos_
    print(f"{word} ({pos_tag}): {freq}")

print('\nСамые часто встречающиеся слова в тексте "Мороз, Красный нос":')
for word, freq in word_freq2.most_common(5):
    token = nlp(word)[0]
    pos_tag = token.pos_
    print(f"{word} ({pos_tag}): {freq}")

print('\nСамые часто встречающиеся слова в тексте "Железная дорога":')
for word, freq in word_freq3.most_common(5):
    token = nlp(word)[0]
    pos_tag = token.pos_
    print(f"{word} ({pos_tag}): {freq}")

print('\nСамые часто встречающиеся слова в тексте "Дедушка Мазай и зайцы":')
for word, freq in word_freq4.most_common(5):
    token = nlp(word)[0]
    pos_tag = token.pos_
    print(f"{word} ({pos_tag}): {freq}")

print('\nСамые часто встречающиеся слова в тексте "Русские женщины":')
for word, freq in word_freq5.most_common(20):
    token = nlp(word)[0]
    pos_tag = token.pos_
    print(f"{word} ({pos_tag}): {freq}")

print('\nСамые часто встречающиеся слова в тексте "Кому на Руси жить хорошо":')
for word, freq in word_freq6.most_common(20):
    token = nlp(word)[0]
    pos_tag = token.pos_
    print(f"{word} ({pos_tag}): {freq}")

# ADJ: Прилагательное (Adjective)
# ADP: Предлог (Adposition)
# ADV: Наречие (Adverb)
# AUX: Вспомогательный глагол (Auxiliary verb)
# CONJ: Союз (Conjunction)
# DET: Определитель (Determiner)
# INTJ: Междометие (Interjection)
# NOUN: Существительное (Noun)
# NUM: Числительное (Numeral)
# PART: Частица (Particle)
# PRON: Местоимение (Pronoun)
# PROPN: Имя собственное (Proper noun)
# PUNCT: Пунктуация (Punctuation)
# SCONJ: Подчинительный союз (Subordinating conjunction)
# SYM: Символ (Symbol)
# VERB: Глагол (Verb)
# X: Другое (Other)

Самые часто встречающиеся слова в тексте "Крестьянкие дети":
поле (NOUN): 5
большая (ADJ): 5
было (AUX): 5
мои (DET): 4
щели (NOUN): 4

Самые часто встречающиеся слова в тексте "Мороз, Красный нос":
Но (CCONJ): 17
бы (AUX): 14
был (AUX): 9
то (SCONJ): 9
солнце (NOUN): 8

Самые часто встречающиеся слова в тексте "Железная дорога":
народ (NOUN): 6
эту (DET): 5
дорогу (NOUN): 4
Всё (NOUN): 3
этот (DET): 3

Самые часто встречающиеся слова в тексте "Дедушка Мазай и зайцы":
Мазай (PROPN): 8
бы (AUX): 4
Мазая (PROPN): 3
мои (DET): 3
Дедушка (NOUN): 2

Самые часто встречающиеся слова в тексте "Русские женщины":
Но (CCONJ): 72
мой (DET): 34
Княгиня (NOUN): 33
был (AUX): 33
было (AUX): 25
его (PRON): 23
была (AUX): 22
то (SCONJ): 19
Губернатор (NOUN): 19
путь (NOUN): 18
всJ (NOUN): 18
нет (VERB): 17
моя (DET): 17
люди (NOUN): 17
сказал (VERB): 15
были (AUX): 15
отец (NOUN): 14
бы (AUX): 14
сам (ADJ): 14
Отец (NOUN): 14

Самые часто встречающиеся слова в тексте "Кому на Руси жить хорошо":
бы (AUX

## Вывод.

**Для проведения морфологического анализа я взял творчество `Николая Алексеевича Некрасова` 1860-1870х годов. Взятые мной произведения Некрасова являются `стихотворениями и поэмами`.**

**Первым делом я определил `лексическое разнообразие произведений`. Примечательно, что если брать произведения писателя в хронологическом порядке, то `наиболее` лексически разнообразными являются те, которые приходятся на промежуток времени с 1865-1871 годы (макс. значения: `0.72` и `0.76`).**

**Можно предположить, что `самым красочным` из выбранных произведений Некрасова является "Железная дорога". Мое предположение следует из того, что именно в этом произведении писатель использует наибольшее количество уникальных прилагательных, где значение лексического разнообразия прилагательных равняется `0.235.`**

**Также, я проанализировал какие слова чаще всего встречаются в выбранных произведениях. Очевидно, что самыми часто встречающимися словами будут являться союзы, наречия, частицы и т.п., так как они в большинстве своем являются связующими звеньями при составлении предложений. Поэтому я сделал дополнительную проверку, чтобы исключить ненужные мне для анализа слова. Из получившегося списка можно сделать вывод, что Некрасов часто упоминал слова, связанные с `окружающей средой`: "Солнце", "Поле", "Дорога". Помимо этого, можно отметить, что писатель не прочь был затронуть и `социальную среду`, так как часто всплывают слова как: "Народ", "Крестьяне", "Странники", "Бог", "Люди", "Губернатор", "Княгиня".**

**Но в целом сложно сказать как менялось творчество писателя по мере его становления из выбранных мной произведений. Я смог использовать в своих целях только самые популярные, но они пришлись на зрелость писателя и на довольно короткий интервал из его творчества. Для лучшего анализа можно было бы взять произведения, написанные Некрасовым в более раннем возрасте, которые в свою очередь затрагивали бы другие времена, где могли бы подниматься другие темы и проблемы, но у меня не получилось их использовать для работы.**